src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala

```diff
@@ -0,0 +1,205 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.commons.io
+
+import java.nio.file.Path
+
+import com.fulcrumgenomics.commons.util.LazyLogging
+
+import scala.collection.mutable.ListBuffer
+import scala.util.Try
+
+trait CommandLineTool extends LazyLogging {
+
+  /** The name of the executable such as Rscript or gs. */
+  val Executable: String
+
+  /** Command used for the above executable.
+    *
+    * For example, to run `Rscript --version`:
+    * {{{
+    * scala> import com.fulcrumgenomics.commons.io._
+    * scala> val TestArgs = Seq("--version")
+    * scala> val Executable = "Rscript"
+    * scala> Rscript.execArgs(Executable +: TestArgs:_*)
+    * }}}
+    * */
+  val TestArgs: Seq[String]
+
+  /** Exception class that holds onto an executable's exit/status code. */
+  case class ToolException(status: Int) extends RuntimeException {
+    override def getMessage: String = s"$Executable failed with exit code $status."
+  }
+
+  /** Executes a command and returns the stdout.
+    *
+    * @param command the command to be executed
+    */
+  def execArgs(command: String*): Try[ListBuffer[String]] = {
```

Comment:
Name of function does not suit action of function. See difference in word "args" vs "command".

I would have expected to use this like:

```scala
Python3.execArgs(Python3.Executable, "-m", "json.tool")
```

If we keep the signature, the function name should be `execCommand`.

Reply:
So syntax would better be:

```scala
Python3.execCommand(Python3.Executable, "-m", "json.tool")
```

However, I would have expected a public definition like:

```scala
Python3.execArgs("-m", "json.tool")
```

Reply:
Consider too:

```scala
protected def execArgs(args: String*): Try[ListBuffer[String]] = {
```

Reply:
When limiting it to `private` it cannot be accessed by `Rscript`:
```
[error] /Users/fangyinlo/fyl_repos/fulcrum-commons/src/test/scala/com/fulcrumgenomics/commons/io/CommandlineToolTest.scala:22:21: method execCommand in trait CommandLineTool cannot be accessed in object com.fulcrumgenomics.commons.io.Rscript
```
Also, `sealed` only applies to `class`

Reply:
Right, private would limit access to the object in which it was defined. Protected, however, would limit access to users of the object, but the self-class and sub-classes could use the member directly.

As for traits, sealing is common in Scala, see: https://www.geeksforgeeks.org/scala-sealed-trait/

Reply:
> Protected, however, would limit access to users of the object, but the self-class and sub-classes could use the member directly.

Sorry, I meant to say **protected**. The protected method cannot be accessed by `Rscript` here: https://github.com/fangylo/commons/blob/ec39999936b121365c212eec51227e8ae11c9488/src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala#L185

The above was the error from protected.

I also meant to say I can't to `seal` on this method specifically

Reply:
However if change to protected it still _can_ access by trait `Versioned` and `Modular` such as https://github.com/fangylo/commons/blob/ec39999936b121365c212eec51227e8ae11c9488/src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala#L159 and https://github.com/fangylo/commons/blob/ec39999936b121365c212eec51227e8ae11c9488/src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala#L172

(It cannot if it were `private`)

It's the `object Rscript` that cannot access the method

Reply:
Ok that makes sense to me, because we are self-typing and not inheriting.